stdlib/public/core/Slab.swift

```diff
@@ -0,0 +1,479 @@
+  /// passed in first element.
+  ///
+  /// This will call the closure 'count' times, where 'count' is the static
+  /// count of the vector, to initialize every element by passing the closure
+  /// an immutable borrow reference to the first element given to the
+  /// initializer. The closure is allowed to throw an error at any point during
+  /// initialization at which point the vector will stop initialization,
+  /// deinitialize every currently initialized element, and throw the given
+  /// error back out to the caller.
+  ///
+  /// - Parameter first: The first value to insert into the vector which will be
+  ///                    passed to the closure as a borrow.
+  /// - Parameter next: A closure that passes in an immutable borrow reference
+  ///                   of the given first element of the vector which returns
+  ///                   an owned `Element` instance to insert into the vector.
+  @available(SwiftStdlib 6.1, *)
+  @_alwaysEmitIntoClient
+  public init<E: Error>(
+    first: consuming Element,
+    next: (borrowing Element) throws(E) -> Element
+  ) throws(E) {
+    // FIXME: We should be able to mark 'Builtin.emplace' as '@once' or something
+    //        to give the compiler enough information to know we will only run
+    //        it once so it can consume the capture. For now, we use an optional
+    //        and take the underlying value within the closure.
+    var o: Element? = first
+
+    self = try Builtin.emplace { (rawPtr) throws(E) -> () in
+      let buffer = Slab<count, Element>._initializationBuffer(start: rawPtr)
+
+      buffer.initializeElement(at: 0, to: o.take()._consumingUncheckedUnwrapped())
+
+      for i in 1 ..< count {
+        do throws(E) {
+          try buffer.initializeElement(at: i, to: next(buffer[i &- 1]))
+        } catch {
+          // The closure threw an error. We need to deinitialize every element
+          // we've initialized up to this point.
+          for j in 0 ..< i {
+            buffer.deinitializeElement(at: j)
+          }
+
+          throw error
+        }
+      }
+    }
+  }
+}
+
+@available(SwiftStdlib 6.1, *)
+extension Slab where Element: Copyable {
+  /// Initializes every element in this vector to a copy of the given value.
+  ///
+  /// - Parameter value: The instance to initialize this vector with.
+  @available(SwiftStdlib 6.1, *)
+  @_alwaysEmitIntoClient
+  public init(repeating value: Element) {
+    self = Builtin.emplace {
+      let buffer = Slab<count, Element>._initializationBuffer(start: $0)
+
+      buffer.initialize(repeating: value)
+    }
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// Collection APIs
+//===----------------------------------------------------------------------===//
+
+@available(SwiftStdlib 6.1, *)
+extension Slab where Element: ~Copyable {
+  /// A type representing the collection's elements.
```

Comment:
Weak suggestion:
```suggestion
  /// The type of the container's elements.
```

Reply:
I simply copied Collection's Element documentation, we can change it if you think we should.

Reply:
The doc for the `Element` type alias has always struck me as oddly wishy-washy. Maybe we can defer a bunch of these doc-comment changes for when we have the containers documentation.